### ISSUE TITLE
Add a database

### ISSUE THREAD

**Issue by mostaphaRoudsari** [ladybug-tools/honeybee-server-daylight#4]
I'm still not sure what is the best approach here. Here are a number of questions:

1. Should we use a SQL database or MangoDB?
2. What is the best approach to dump Radiance's results in a database? What are the values to keep track of? @rpg777 I think I saw that you tried this for OpenStudio at some point but I can't find it now. Thoughts? Links?
3. How does the database integrates with individual tasks? Ideally each run will be broken in multiple tasks (e.g. total-sky, direct-sky and direct-sun). Should them be dumped directly into the database or should I wait for all of them to be finished and then dump the final results to the database?
4. ...

**Comment by rpg777** [comment]
Perhaps the first question is why do you feel you even need a results database? 

But to answer your questions here, yes we dump out the illuminance map data to a mysql database, just sorta following the format of eplusout. With the disappearance of Results Viewer (r.i.p.), we no longer have a built-in method fo visualize that data, but that data is so limited anyway. We need a new solution. We have similar [issue](https://github.com/NREL/OpenStudio/issues/1124) in OpenStudio and I was actually going to talk to you about this, @mostaphaRoudsari! We need better visualizations, and definitely better methods for storing the data. 

We use MongoDB in OpenStudio Server, and are starting to use PostgreSQL on some projects; you'd have to talk to Nick Long about that. 

Q3 speaks to the intractable nature of the multi-phase daylight simulation methodologies, especially when applied to multiple window groups. You end up with so much data! The value is in being able to do all kinds of what-if analyses and optimization of control strategies, but that requires retention of all the individual datasets, and storing and working with these sets can be very difficult. I think that clever use of databases IS the key to all of this, but we haven't really started rolling up our sleeves on this end to tackle this. Maybe we can work together.

**Comment by mostaphaRoudsari** [comment]
> Perhaps the first question is why do you feel you even need a results database?

Read here: https://github.com/ladybug-tools/honeybee/issues/195
Basically for the same reasons that you noted in Q3. Honeybee does handle the what if scenarios by loading all the results for each sensor. It works fine from command line but from inside Grasshopper it slows down the interface and that's why I started thinking about the database.

Also when it comes to scaling the analysis for parallel runs I'm thinking to break down each phase of the analysis as a separate task and run it in parallel which can end up on a different worker. Right now I mount a directory so they all write the results into the same folder but ideally we want to push all of them directly to the database and do all the post-processing inside the database which should be faster than radiance anyways.  

@nllong, hi!, What would be your input on database. In structured databases sending queries for daylight metrics is pretty straight forward. I haven't used MongoDB so I don't know the dis/advantages. 

@rpg777, while waiting for @nllong I think I go ahead and make a first iteration using sqlite. Once we have a prototype it should be a good time to set up a call and chat unless you want to talk before that.

**Comment by AntoineDao** [comment]
Personal opinion on this one would be to go for MongoDB for two main reasons:

1. Flexible Schema: the schema set up is much more flexible than SQL which means you can easily start grouping different types of analysis in the future for a similar design state
2. Bson ~= Json: it seems honeybee is moving towards Json formats to transfer information across services. MongoDB formats everything in Json which means there will me minimal translation work to transform HB json objects/results into a database. You could even have set ups whereby you send all of your design states in Json format to your MongoDB and later on process all of them through HPC clusters mounting the database.

**Comment by mostaphaRoudsari** [comment]
Hi @AntoineDao,

I was also in favor of MongoDB as the current structure of HB[+] for inputs and outputs are based on dictionaries which is very easy to translate to JSON but unfortunately [we won't be able to use pyMongo in IronPython](https://stackoverflow.com/a/14608515/4394669). Also after spending some time thinking about the structure of outputs it makes sense to use a SQL database for outputs. It takes rewriting both `AnalysisPoint` and `AnalysisGrid` classes and updating all the post-processing components but it does worth the effort.

Right now I'm thinking to use postgreSQL for docker image which does handle both SQL and JSON data types to store the input JSONS and output results, and use sqlite for storing the results on desktop. I will post the current database design that I came up with under its own issue: https://github.com/ladybug-tools/honeybee/issues/195

**Comment by nllong** [comment]
I second the use of PostgreSQL and the JSON field type would be the way to
go. It provides the best of both worlds.

**Comment by mostaphaRoudsari** [comment]
Thanks @nllong for the input! I hopefully get the implementation started at some point this week.

**Comment by nllong** [comment]
no problem. let me know if have any postgres questions.

**Comment by arif-hanif** [comment]
you really have relational data mostapha

**Comment by AntoineDao** [comment]
Oh boy rereading my comment to this is making me pale in shame. I don't think mongo is a good shout anymore 😛 Is there anything blocking you from merging this in @mostaphaRoudsari ?

Happy to help if needed, however it's looking like this PR is OOOOOLD so it's unlikely to play nice with current code base. Maybe worth putting this as a hacktoberfest issue? Maybe https://github.com/ladybug-tools/honeybee/issues/195?

**Comment by mostaphaRoudsari** [comment]
I'm on the edge to delete this repository or make it read-only. With all that we have learned during the last year this app can be considered as very inefficient. It will be a totally different app if we want to write it now. I think it is better to just replace this app with a newer version once we are happy with the work that is happening for Pollination. And yes, Postgres will be my preferred option. I had a very good experience using it with https://github.com/MagicStack/asyncpg
